archive/frames/views.py

```diff
@@ -422,6 +425,81 @@ def get_endpoint_name(self):
+        """
+        Filter thumbnails depending on the logged in user.
+        Admin users see all thumbnails
+        Authenticated users see all frames with a PUBDAT in the past, plus
+        all thumbnails that belong to frames under their proposals.
+        Non authenticated see all thumbnails for frames with a PUBDAT in the past
+        """
+        queryset = (
+            Thumbnail.objects.all().select_related('frame')
+        )
+        if self.request.user.is_superuser:
+            return queryset
+        elif self.request.user.is_authenticated:
+            return queryset.filter(
+                Q(frame__proposal_id__in=self.request.user.profile.proposals) |
+                Q(frame__public_date__lt=datetime.datetime.now(datetime.timezone.utc))
+            )
+        else:
+            return queryset.filter(frame__public_date__lt=datetime.datetime.now(datetime.timezone.utc))
+
+    # These two method overrides just force the use of the as_dict method for serialization for list and detail endpoints
+    def list(self, request, *args, **kwargs):
+        queryset = self.filter_queryset(self.get_queryset())
+        page = self.paginate_queryset(queryset)
+        json_models = [model.as_dict() for model in page]
+        return self.get_paginated_response(json_models)
+
+    def retrieve(self, request, *args, **kwargs):
+        instance = self.get_object()
+        return Response(instance.as_dict())
+
+    def create(self, request):
+        basename = request.data.get('basename')
+        logger_tags = {'tags': {
+            'filename': basename,
+            'request_id': request.data.get('request_id')
+        }}
+        logger.info('Got request to process thumbnail', extra=logger_tags)
+        # Make sure we have the minimum information to make a frame object associated with the thumbnail if one doesn't already exist
+        frame_serializer = FrameSerializer(data=request.data)
+        if frame_serializer.is_valid():
+            # The ingester sends the key/extension in the version_set, but we don't keep versions of thumbnails, so just store the key/extension in the thumbnail
+            request.data['key'] = request.data['version_set'][0]['key']
+            request.data['extension'] = request.data['version_set'][0]['extension']
+        else:
+            logger_tags['tags']['errors'] = frame_serializer.errors
+            logger.fatal('Request to process thumbnail failed when serializing frame metadata', extra=logger_tags)
+            return Response(frame_serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        thumbnail_serializer = ThumbnailSerializer(data=request.data)
+        if thumbnail_serializer.is_valid():
+            # Remove the version set as this version does not correspond to the frame object, but rather the thumbnail.
+            del frame_serializer.validated_data['version_set']
+            frame = frame_serializer.save(basename=request.data['frame_basename'])
```

Comment:
Don't we want to just get the existing frame if it exists, or save/create it if it doesn't?